### DISCUSSION TITLE
rke2-agent node is not able to join to rke2-server cluster due to error: failed to get CA certs

### DISCUSSION THREAD

**Discussion by anishagoyal0504** [rancher/rke2#6532] [General]
Hi Team,
My rke2-server is working fine and can see the rke2-server node in the cluster (kubectl get nodes).
Now I am trying to join rke2-agent node in the rke2-server cluster by starting the service however, it is failing on the below:

"failed to get CA certs: Get \"https://127.0.0.1:6444/cacerts\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"

my rke2-agent node has below config to join the rke2-server clsuter:
```yaml
server: https://<rke2-server private ip address>:9345
token: <value of node-token>
```

Am I missing any steps for installing rke2-server and rke-agent. I am using standard method for installation (curl -sfL https://get.rke2.io | INSTALL_RKE2_TYPE="agent" sh -)

Any clue, why this error is coming, I could not find any solid answer or solution either on google or rke2- troubleshooting guides.

Any view please

thanks

**Comment by brandond** [comment]
It looks like the agent can't connect to the server? The connection is timing out. Make sure that you can successfully curl the server address from the agent.

**Reply by anishagoyal0504** [reply]
I can curl the rke2-server ip from rke2-agent as per below
$ curl 10.119.80.119
<html>
<head><title>404 Not Found</title></head>
<body>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx</center>
</body>
</html>

**Reply by brandond** [reply]
Wrong port and protocol, you're curling the nginx ingress, not rke2. Use the server URL as you have it specified in the config file.

**Reply by anishagoyal0504** [reply]
I do not know why you are relating to rke2 server ip address to nginx ingress, can you please review your comment agan, what triggers you it is related to nginx ingress, this is my config file on rke2-agent AWS VM:

sh-4.4$ sudo cat /etc/rancher/rke2/config.yaml
server: https://10.119.80.119:9345

I curl the above but nothing coming back. the above IP is my rke2-server private IP address on AWS linux VM.

I

**Reply by brandond** [reply]
You didn't curl `https://10.119.80.119:9345`, you curled `10.119.80.119` which is equivalent to `http://10.119.80.119:80` and would be served by the ingress controller, not the rke2 supervisor.  You need to actually curl the correct things when testing.

If you try to curl the correct address and nothing comes back, that would be your problem. Check the docs and make sure you've opened the correct ports between nodes.

**Reply by anishagoyal0504** [reply]
wondering if you have tried installing RKE2 on AWS linux VM, 

after fixing inbound security rule son aws vm, now I am getting this error on rke2-agent
=error msg="Error encountered while importing /var/lib/rancher/rke2/agent/images/runtime-image..txt: image \"index.doc>>

**Reply by brandond** [reply]
> after fixing inbound security rules

Ah, so you were indeed  blocking critical ports.

> `Error encountered while importing`

You truncated that error message so I have no idea what the actual error is. You could try attaching the full log from journald. Some errors and warnings during startup are normal, but I can't tell based on part of a single message.

**Reply by anishagoyal0504** [reply]
```console
sh-5.1$ sudo journalctl -u rke2-agent -xe
 systemd[1]: Stopped Rancher Kubernetes Engine v2 (agent).
░░ Subject: A stop job for unit rke2-agent.service has finished
░░ Defined-By: systemd
░░ Support: https://access.redhat.com/support
░░
░░ A stop job for unit rke2-agent.service has finished.
░░
░░ The job identifier is 7951 and the job result is done.
 systemd[1]: rke2-agent.service: Consumed 10.667s CPU time.
░░ Subject: Resources consumed by unit runtime
░░ Defined-By: systemd
░░ Support: https://access.redhat.com/support
░░
░░ The unit rke2-agent.service completed and consumed the indicated resources.
 systemd[1]: Starting Rancher Kubernetes Engine v2 (agent)...
░░ Subject: A start job for unit rke2-agent.service has begun execution
░░ Defined-By: systemd
░░ Support: https://access.redhat.com/support
░░
░░ A start job for unit rke2-agent.service has begun execution.
░░
░░ The job identifier is 7951.
 sh[45300]: + /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service
 modprobe[45302]: libkmod: kmod_config_parse: /etc/modprobe.d/sctp.conf line 1: ignoring bad line starting with 'install'
 modprobe[45303]: libkmod: kmod_config_parse: /etc/modprobe.d/sctp.conf line 1: ignoring bad line starting with 'install'
 rke2[45304]: time="2024-08-16T11:28:55+10:00" level=warning msg="not running in CIS mode"
 rke2[45304]: time="2024-08-16T11:28:55+10:00" level=info msg="Applying Pod Security Admission Configuration"
 rke2[45304]: time="2024-08-16T11:28:55+10:00" level=info msg="Starting rke2 agent v1.28.12+rke2r1 (27989db0c83739c9f560e0567a3bf76177947122)"
 rke2[45304]: time="2024-08-16T11:28:55+10:00" level=info msg="Adding server to load balancer rke2-agent-load-balancer: 10.119.80.119:9345"
 rke2[45304]: time="2024-08-16T11:28:55+10:00" level=info msg="Running load balancer rke2-agent-load-balancer 127.0.0.1:6444 -> [10.119.80.119:93>
 rke2[45304]: time="2024-08-16T11:28:55+10:00" level=info msg="Adding server to load balancer rke2-api-server-agent-load-balancer: 10.119.80.119:>
 rke2[45304]: time="2024-08-16T11:28:55+10:00" level=info msg="Running load balancer rke2-api-server-agent-load-balancer 127.0.0.1:6443 -> [10.11>
 rke2[45304]: time="2024-08-16T11:28:56+10:00" level=info msg="Module overlay was already loaded"
 rke2[45304]: time="2024-08-16T11:28:56+10:00" level=info msg="Module nf_conntrack was already loaded"
 rke2[45304]: time="2024-08-16T11:28:56+10:00" level=info msg="Module br_netfilter was already loaded"
 rke2[45304]: time="2024-08-16T11:28:56+10:00" level=info msg="Module iptable_nat was already loaded"
 rke2[45304]: time="2024-08-16T11:28:56+10:00" level=info msg="Module iptable_filter was already loaded"
 rke2[45304]: time="2024-08-16T11:28:56+10:00" level=info msg="Runtime image index.docker.io/rancher/rke2-runtime:v1.28.12-rke2r1 bin and charts >
 rke2[45304]: time="2024-08-16T11:28:56+10:00" level=info msg="Removed kube-proxy static pod manifest"
 rke2[45304]: time="2024-08-16T11:28:56+10:00" level=info msg="Logging containerd to /var/lib/rancher/rke2/agent/containerd/containerd.log"
 rke2[45304]: time="2024-08-16T11:28:56+10:00" level=info msg="Running containerd -c /var/lib/rancher/rke2/agent/etc/containerd/config.toml -a /r>
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="containerd is now running"
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="Pulling images from /var/lib/rancher/rke2/agent/images/kube-proxy-image.txt"
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="Image index.docker.io/rancher/hardened-kubernetes:v1.28.12-rke2r1-build20240717 ha>
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="Imported docker.io/rancher/hardened-kubernetes:v1.28.12-rke2r1-build20240717"
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="Imported images from /var/lib/rancher/rke2/agent/images/kube-proxy-image.txt in 6.>
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="Pulling images from /var/lib/rancher/rke2/agent/images/runtime-image.txt"
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="Image index.docker.io/rancher/rke2-runtime:v1.28.12-rke2r1 has already been pulled"
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="Imported docker.io/rancher/rke2-runtime:v1.28.12-rke2r1"
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="Imported images from /var/lib/rancher/rke2/agent/images/runtime-image.txt in 5.727>
 rke2[45304]: time="2024-08-16T11:28:57+10:00" level=info msg="Getting list of apiserver endpoints from server"
 rke2[45304]: time="2024-08-16T11:28:58+10:00" level=info msg="Got apiserver addresses from supervisor: [10.119.80.119:6443]"
 rke2[45304]: time="2024-08-16T11:31:10+10:00" level=info msg="Connecting to proxy" url="wss://10.119.80.119:9345/v1-rke2/connect"
 rke2[45304]: time="2024-08-16T11:31:10+10:00" level=info msg="Creating rke2-cert-monitor event broadcaster"
 rke2[45304]: time="2024-08-16T11:31:10+10:00" level=info msg="Running kubelet --address=0.0.0.0 --allowed-unsafe-sysctls=net.ipv4.ip_forward,net>
 rke2[45304]: time="2024-08-16T11:31:10+10:00" level=info msg="Remotedialer connected to proxy" url="wss://10.119.80.119:9345/v1-rke2/connect"
 rke2[45304]: time="2024-08-16T11:31:10+10:00" level=info msg="Running kube-proxy --cluster-cidr=10.42.0.0/16 --conntrack-max-per-core=0 --conntr>
lines 2022-2078/2078 (END)
```

**Reply by anishagoyal0504** [reply]
I found the issue on AWS linux VM Inbound security rules on RKE2-Server, when I do netstat on rke2-server, it shows the ports are opening on tcp6 (::/0) IPv6, when I added IPv6 (::/0) in inbound security rules, rke2-agent node was able to join the rke2 cluster. I was under the impression that particularly these ports (6443, 9345) are should be running under TCP4 on rke2-server. you might need to add this in RKE2 troubleshooting guidelines so that people knows straightway.

**Reply by anishagoyal0504** [reply]
Hi @brandond I have created another rke2 cluster on aws EC2 instance. I am having same issue with joining rke2-agent to rke2-server, when I start the rke2-agent service it gives me the error: 
failed to get CA certs: Get \"https://127.0.0.1:6444/cacerts\": context deadline exceeded
why this is misleading logs and why it is pointing to port 6444, what to do with this port?

in my rke2-server config.yaml has:
server: https://10.119.61.128:9345
token: K10401aeba15555fd9c124f50f516fe9399b41af273a93469d68077e716d5351de6::server:075ba65d8e7d20e07ecd5555cbf87f25

Can you please me with this and there is known issue on AWS. When I run the same setup on my local desktop machine instead of AWS, it works fine

**Reply by manuelbuil** [reply]
6444 is the proxy port. It just makes sense locally. That will forward the connection to the server, if you check the logs, you'll see something like this: 
```
 rke2[45304]: time="2024-08-16T11:28:55+10:00" level=info msg="Running load balancer rke2-agent-load-balancer 127.0.0.1:6444 -> $IP:$PORT
```
Please make sure your AWS security group has $PORT open

**Reply by anishagoyal0504** [reply]
Hi @manuelbuil , my aws security group has all the ports (9345,6443,2379,10250, 6444) are open in inbound security rules. I have no idea why it is still causing this issue?

running netstat on rke2-server:
[ec2-user@ip-10-119-61-128 ~]$ netstat -aln | grep 9345
tcp        0      0 10.119.61.128:40988     10.119.61.128:9345      ESTABLISHED
tcp        0      0 127.0.0.1:37442         127.0.0.1:9345          ESTABLISHED
tcp6       0      0 :::9345                 :::*                    LISTEN     
tcp6       0      0 127.0.0.1:9345          127.0.0.1:37442         ESTABLISHED
tcp6       0      0 10.119.61.128:9345      10.119.61.128:40988     ESTABLISHED

when I trying to curl from rke2-agent:
curl -vks https://10.119.61.128:9345/cacerts

nothing happened

looking at logs of rke2-agent node it says:
"failed to get CA certs: Get \"https://127.0.0.1:6444/cacerts\": context deadline ex

It looks like it is known issue on AWS based VM, it should be documented in your rancher guidlines

Can you help please

**Reply by brandond** [reply]
If you can't curl the server URL then there is some other issue at the network level. There are no known issues with running rke2 on EC2. You just need to make sure that the nodes can reach each other, same as in any other environment.
